### PR TITLE
Don't send follow up notifications to newly removed collabs

### PR DESCRIPTION
**Pull request by jsncmgs1**
## Context

Currently we send follow up messages to everyone who was on the original message, even if that collaborator has since been removed from the app. This fixes that by checking collaborators on the message against what API says current collaborators are.

I talked to Sepi about how this should work - if any of the original collabs exist, only send it to them. If none of the original collabs exist, send to all new collaborators. Details in the GUS chatter feed.

## SOC2
https://gus.my.salesforce.com/a07B00000077AJJIA2

### CODE REVIEW

File: lib/mediators/followups/notifier.rb [hunk 1]
```diff
@@ -14,10 +14,23 @@ def call
 
     private
 
+    def heroku_client
+      @client ||= Telex::HerokuClient.new
+    end
+
+    def filter_notifications
```
Comment (edmorley) [inline lib/mediators/followups/notifier.rb:21]: Hi! If I'm understanding this change correctly it means that for follow-up emails, the notifications list will now only include users who are both a current collaborator, and also were a collaborator at the time of the earlier notification for which this is a follow-up?

If so, then that means if a collaborator was added to an app after the initial notification, then they won't receive the follow-ups either?

It seems like follow-ups should ignore the earlier notifications list entirely, and use only the collaborator list from API, so that all current collaborators are emailed?

Comment (jsncmgs1) [reply lib/mediators/followups/notifier.rb:21]: @edmorley Good question. Now that you mentioned that, something I'm unsure of is if a collaborator has joined after the original message was sent out, should they get the follow up to the original message? If so, then yes, you are correct, this needs updating. I was operating under the assumption that new collabs would not receive followups for old messages

Comment (jsncmgs1) [reply lib/mediators/followups/notifier.rb:21]: Thinking about it more, the current code makes no attempt to add new collabs to the notifications list, but since the owner is considered a collab in this context, someone always gets the notification. Seems like the original intent was to limit it to those who received the original message

Comment (mikehale) [reply lib/mediators/followups/notifier.rb:21]: I think it might be better to send to the full list of current collabs whether or not they were on the original message list. Any new collabs will not see the original context, but some context is likely better than no context.

Here is an example of a notification with a followup:

image

All the folllowups will share the [subject with the original message](https://github.com/heroku/telex/blob/7ab389934a920321137f9e8b2afc53a974a49ef4/lib/mediators/followups/notifier.rb#L30), so that could also provide a valuable clue for new collabs to chase down critical issues.

**Comment by jsncmgs1** [comment]
Looping a PM into the convo.

File: spec/mediators/followups/notification_updater_spec.rb [hunk 2]
```diff
@@ -0,0 +1,37 @@
+RSpec.describe Mediators::Followups::NotificationUpdater do
+  include HerokuAPIMock
+
+  before do
```
Comment (jsncmgs1) [inline spec/mediators/followups/notification_updater_spec.rb:4]: I'm not a fan if ivars here but I'm trying to keep it consistent with the other tests I've been working in.

**Comment by jsncmgs1** [comment]
Update from Sepi oh how this should work in the gus ticket:

TL;DR, if any of the original collabs exist, only send it to them. If none of the original collabs exist, send to all.

File: lib/mediators/followups/notification_updater.rb [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+    def update_notifications
+      if notifications_for_collabs_that_recieved_original_message.any?
+        notifications_for_collabs_that_recieved_original_message
+      else
+        create_notifications_for_new_collabs
+      end
+    end
+
+    def notifications_for_collabs_that_recieved_original_message
+      message.notifications.select do |n|
```
Comment (jsncmgs1) [inline lib/mediators/followups/notification_updater.rb:27]: Should be destroying old notifications here instead of filtering? Maybe, but I don't know telex well enough to be comfortable with that at the moment.

File: lib/mediators/followups/notification_updater.rb [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+    end
+
+    private
+
+    # if there exists collaborators who received the original message, send it to them.
+    # otherwise, send it to all collabs. Sometimes this user may not exist in telex yet,
+    # so we have to create them.
+    def update_notifications
+      if notifications_for_collabs_that_recieved_original_message.any?
+        notifications_for_collabs_that_recieved_original_message
```
Comment (jessieay) [inline lib/mediators/followups/notification_updater.rb:20]: maybe to reveal intent we can rename this `current_collabs_that_recieved_original_message` or something like that? Because this is not *just* those collabs, it's those collabs with removed collabs removed from list

File: lib/mediators/followups/notification_updater.rb [hunk 5]
```diff
@@ -0,0 +1,60 @@
...
+
+    def call
+      update_notifications
+    end
+
+    private
+
+    # if there exists collaborators who received the original message, send it to them.
+    # otherwise, send it to all collabs. Sometimes this user may not exist in telex yet,
+    # so we have to create them.
```
Comment (jessieay) [inline lib/mediators/followups/notification_updater.rb:17]:
```suggestion
    # if any collaborators who received the original message are still collabs, send it to just that list minus 
    # anyone who was removed as a collab. Otherwise, send it to all current collabs. 
    # Sometimes a user may not exist in telex yet, so have to create them.
```

File: lib/mediators/followups/notification_updater.rb (hunk 3 above)
Comment (jessieay) [reply lib/mediators/followups/notification_updater.rb:27]: is this question unanswered or was it answered elsewhere? would be good to know answer :)

File: lib/mediators/followups/notification_updater.rb [hunk 6]
```diff
@@ -0,0 +1,60 @@
...
+      end
+    end
+
+    def notifications_for_collabs_that_recieved_original_message
+      message.notifications.select do |n|
+        current_collab_hids.include?(n.user.heroku_id)
+      end
+    end
+
+    def create_notifications_for_new_collabs
```
Comment (jessieay) [inline lib/mediators/followups/notification_updater.rb:32]: if I am reading this method correctly, I think a better name might be `create_notifications_for_all_collabs` - what do you think?

File: lib/mediators/followups/notification_updater.rb [hunk 7]
```diff
@@ -0,0 +1,60 @@
...
+
+    def current_collab_hids
+      current_collabs.map {|c| c["user"]["id"] }
+    end
+
+    def current_collabs
+      @current_collabs ||= Telex::HerokuClient.new.app_collaborators(message.target_id)
+    end
+  end
+end
```
Comment (jessieay) [inline lib/mediators/followups/notification_updater.rb:60]: no newline at EOF! so annoying that it's not the default setting on all editors

File: spec/mediators/followups/notification_updater_spec.rb (hunk 2 above)
Comment (jessieay) [reply spec/mediators/followups/notification_updater_spec.rb:4]: makes sense.

it seems like most of these are only used in one test, thoughts on putting this setup into just the first test so we aren't doing common setup unnecessarily? 2nd one only needs `@heroku_app ` as far as I can tell

File: spec/mediators/followups/notification_updater_spec.rb [hunk 8]
```diff
@@ -0,0 +1,37 @@
...
+  end
+
+  it "creates and returns notifications for all collabs if none received the original message" do
+    message = Fabricate(:message,
+      title: Faker::Company.bs,
+      notifications: [Fabricate(:notification)],
+      target_id: @heroku_app.id)
+
+    followup = double("followup", body: Faker::Company.bs, message: message)
+    user3, user4 = Fabricate(:user), Fabricate(:user)
```
Comment (jessieay) [inline spec/mediators/followups/notification_updater_spec.rb:29]: wow Fabricate!~

File: spec/support/mock_heroku_api.rb [hunk 9]
```diff
@@ -59,6 +59,11 @@ def create_heroku_app(owner:, collaborators: [])
     app
   end
 
+  # overrides previously declared stubs by recreating the app using the same app id
```
Comment (jessieay) [inline spec/support/mock_heroku_api.rb:62]: not sure I understand this comment. possible to clarify?

File: lib/mediators/followups/notification_updater.rb (hunk 4 above)
Comment (jsncmgs1) [reply lib/mediators/followups/notification_updater.rb:20]: I struggled with naming on this one, I actually started with something nearly identical to what you suggested but changed it because the return value is a collection of notifications, but the naming suggests it would be the collaborators. Maybe `original_message_recipient_notifications` or something like that?

**Comment by jsncmgs1** [comment]
> it seems like most of these are only used in one test, thoughts on putting this setup into just the first test so we aren't doing common setup unnecessarily? 2nd one only needs @heroku_app as far as I can tell

@jessieay the heroku app needs to know about `@user1` and `@user2`, so we'd have to copy them into both tests. That being said, I prefer minimal setup and am happy to remove it if we want to go that route.

File: lib/mediators/followups/notification_updater.rb (hunk 3 above)
Comment (dmcinnes) [reply lib/mediators/followups/notification_updater.rb:27]: I suspect it's better to leave them as an audit trail. This message was originally sent to these people.

File: spec/mediators/followups/notification_updater_spec.rb (hunk 2 above)
Comment (jsncmgs1) [reply spec/mediators/followups/notification_updater_spec.rb:4]: We need the setup to happen in both cases so that we can assert that @user1 and @user2 are not in the notifications after the collaborator update in the second test.

File: lib/mediators/followups/notification_updater.rb (hunk 3 above)
Comment (jsncmgs1) [reply lib/mediators/followups/notification_updater.rb:27]: I concur.

File: lib/mediators/followups/notification_updater.rb (hunk 6 above)
Comment (jsncmgs1) [reply lib/mediators/followups/notification_updater.rb:32]: I think they're both technically correct since in this scenario all the collabs are new collabs, I'm good with changing it though.

File: spec/support/mock_heroku_api.rb (hunk 9 above)
Comment (jsncmgs1) [reply spec/support/mock_heroku_api.rb]: updated in https://github.com/heroku/telex/pull/125/commits/77d705a6a5bf3ee49322f190a7c3139692371743

File: spec/mediators/followups/notification_updater_spec.rb [hunk 10]
```diff
@@ -0,0 +1,37 @@
+RSpec.describe Mediators::Followups::NotificationUpdater do
+  include HerokuAPIMock
+
+  before do
+    @user1, @user2 = create_heroku_user, create_heroku_user
+    @note1, @note2 = [@user1, @user2].map { |u| double("notification", id: SecureRandom.uuid, user: u) }
+    @heroku_app = create_heroku_app(owner: @user1, collaborators: [@user1, @user2])
```
Comment (jsncmgs1) [inline spec/mediators/followups/notification_updater_spec.rb:7]: We need the setup to run in both tests so we can validate that user1 and user2 will not receive notifications in the second test.

File: spec/mediators/followups/notification_updater_spec.rb [hunk 11]
```diff
@@ -0,0 +1,37 @@
...
+    followup = double("followup", body: Faker::Company.bs, message: message)
+    user3, user4 = Fabricate(:user), Fabricate(:user)
+
+    update_app_collaborators(@heroku_app, [user3, user4])
+    notifications = described_class.run(followup: followup)
+    expected = notifications.map(&:user)
+
+    expect(expected).to contain_exactly(user3, user4)
+  end
+end
```
Comment (kwugirl) [inline spec/mediators/followups/notification_updater_spec.rb:37]: GH wants a newline here.

File: spec/mediators/followups/notifier_spec.rb [hunk 12]
```diff
@@ -1,28 +1,65 @@
...
+
+    user3 = create_heroku_user
+    update_app_collaborators(heroku_app, [user3])
+    described_class.new(followup: followup).call
+
+    ds = Mail::TestMailer.deliveries
+    expect(ds.size).to eq(1)
+    expect(ds.map(&:to).flatten.sort).to eq([user3].map(&:email).sort)
+    expect(ds.map(&:in_reply_to).uniq.sort).to eq(["#{Notification.first.id}@notifications.heroku.com"])
+    end
```
Comment (kwugirl) [inline spec/mediators/followups/notifier_spec.rb:63]: I kind of feel like these scenarios should be covered in the NotificationUpdater specs since that is now responsible for determining that the right people get the follow up emails. So that would mean just adding in a third scenario in that spec for the "some but not all original collabs remain".

File: spec/support/mock_heroku_api.rb [hunk 13]
```diff
@@ -59,6 +59,18 @@ def create_heroku_app(owner:, collaborators: [])
...
   end
 
+  # If you call create_heroku_app with an existing app id and different owners or collabs than
+  # what's already associated with the app, it will rewrite the mocked
+  # responses for that app id to reflect the change in ownership or
+  # collaborators.
+
+  # This method calls create to rewrite the mocked collab response for a given app,
+  # with the goal of being more obvious that an update is happening, rather than
+  # calling create_heroku_app twice in the same test.
```
Comment (kwugirl) [inline spec/support/mock_heroku_api.rb:69]: This is helpful!

File: lib/mediators/followups/notification_updater.rb [hunk 14]
```diff
@@ -0,0 +1,60 @@
+module Mediators::Followups
+  class NotificationUpdater < Mediators::Base
```
Comment (kwugirl) [inline lib/mediators/followups/notification_updater.rb:2]: I feel like this name could possibly be a bit more descriptive... `Followups::NotificationScoper`? `Followsup::RecipientScoper`? I found it slightly confusing to wrap my mind around the idea of updating a notification at first, when I was thinking of notification in the general sense (i.e., "isn't a notification just a notification, why would it be updated?") until I realized that this is more referring to the `Notification` object and updating _its_ properties. I sorted it out in the end, it's not that bad, but wanted to make the suggestion.
